// crossbow/dist/adaptors/@shell.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _npm_1 = require("./@npm");
var task_resolve_1 = require("../task.resolve");
var task_utils_1 = require("../task.utils");
var file = require("../file.utils");
var merge = require("../../lodash.custom").merge;
var debug = require("debug")("cb:@shell");
function default_1(task, trigger) {
    return function (opts, ctx, done) {
        var args = (function () {
            if (task.origin === task_resolve_1.TaskOriginTypes.FileSystem) {
                try {
                    var contentFromDisk = file.readFileContent(task.externalTasks[0]);
                    return _npm_1.getArgs(contentFromDisk);
                }
                catch (e) {
                    return {
                        errors: [e]
                    };
                }
            }
            return _npm_1.getArgs(task.command);
        })();
        if (args.errors.length) {
            done(args.errors[0]);
            return;
        }
        var stdio = _npm_1.getStdio(trigger);
        var cbEnv = task_utils_1.getCBEnv(trigger);
        var ctxEnv = task_utils_1.getContextEnv(trigger, ctx);
        var userEnv = _npm_1.getEnv(process, trigger.config);
        var env = merge({}, process.env, userEnv, cbEnv, task.env, trigger.config.env, ctxEnv);
        debug("running %s", args.cmd);
        var emitter = _npm_1.runCommand(args.cmd, {
            cwd: trigger.config.cwd,
            env: env,
            stdio: stdio
        });
        _npm_1.handleExit(emitter, done);
        return function tearDownShellAdaptor() {
            _npm_1.teardown(emitter, task);
        };
    };
}
exports.default = default_1;
;
