// crossbow/dist/setup.bin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var file_utils_1 = require("./file.utils");
var reports = require("./reporter.resolve");
var task_utils_1 = require("./task.utils");
var _ = require("../lodash.custom");
var fs_1 = require("fs");
var fs_2 = require("fs");
var path_1 = require("path");
var fs_3 = require("fs");
/**
 * If the user has specified the `bin` option,
 * we need to ensure these paths are valid directories
 * and read the executables from them
 */
exports.addBinLookupsToObject = function (config) {
    return config.bin.length
        // take the `config.bin` array and resolve each
        ? exports.getBins(config.bin, config.cwd)
            // if successful, add the directories and executables
            .map(function (bindirs) {
            return _.assign({}, config, {
                binDirectories: bindirs.valid,
                binExecutables: exports.getExecutables(bindirs.valid)
            });
        })
        : file_utils_1.Right(config);
};
/**
 * Resolve many bin paths
 */
exports.getBinLookups = function (paths, cwd) {
    return file_utils_1.Right([].concat(paths).map(function (path) { return exports.getBinLookup(path, cwd); }))
        .chain(function (xs) {
        return {
            all: xs,
            valid: xs.filter(function (x) { return x.errors.length === 0; }),
            invalid: xs.filter(function (x) { return x.errors.length > 0; })
        };
    });
};
/**
 * Resolve a single path. Allow for errors such as
 * not found or not a directory
 */
exports.getBinLookup = function (path, cwd) {
    return joinPath(String(path), cwd)
        .chain(function (resolved) { return file_utils_1.Right(resolved)
        .chain(function (resolved) { return binDirectoryExists(resolved); })
        .chain(function (resolved) { return isDirectory(resolved); })
        .fold(function (error) {
        return {
            errors: [error],
            resolved: resolved,
            input: path
        };
    }, function (resolved) {
        return {
            errors: [],
            resolved: resolved,
            input: path
        };
    }); });
};
/**
 * Resolve many bin directories, but if any produce an
 * error, return a Left() indicating that
 */
exports.getBins = function (dir, cwd) {
    return file_utils_1.Right(exports.getBinLookups(dir, cwd))
        .chain(function (x) { return x.invalid.length
        ? file_utils_1.Left({ type: reports.ReportTypes.BinOptionError, data: x })
        : file_utils_1.Right(x); });
};
/**
 * Scan a directory to retrieve an array of executables
 */
exports.getExecutables = function (dirs) {
    return dirs.reduce(function (acc, lookup) {
        var items = fs_2.readdirSync(lookup.resolved);
        return acc.concat(items.filter(function (dir) {
            try {
                return fs_1.statSync(path_1.join(lookup.resolved, dir)).isFile();
            }
            catch (e) {
                return false;
            }
        }));
    }, []);
};
var binDirectoryExists = function (path) {
    return fs_3.existsSync(path)
        ? file_utils_1.Right(path)
        : file_utils_1.Left({ type: task_utils_1.InputErrorTypes.BinDirectoryNotFound });
};
var isDirectory = function (path) {
    return fs_1.statSync(path).isDirectory()
        ? file_utils_1.Right(path)
        : file_utils_1.Left({ type: task_utils_1.InputErrorTypes.BinPathNotADirectory });
};
var joinPath = function (path, cwd) { return file_utils_1.Right(path_1.join(cwd, path)); };
