// crossbow/dist/setup.envFile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var file_utils_1 = require("./file.utils");
var reports = require("./reporter.resolve");
var task_utils_1 = require("./task.utils");
var _ = require("../lodash.custom");
/**
 * If the user has specified any `envFile` options,
 * we need to read each file from disk and parse it's data.
 * Finally we save the results on the incoming config object
 */
exports.addEnvFilesToObject = function (config) {
    return config.envFile.length
        // Take the `config.envFile` array and resolve each
        ? exports.getEnvFiles(config.envFile, config.envFilePrefix, config.cwd)
            // Add `envFiles` to config object
            .map(function (envFiles) {
            return _.assign({}, config, {
                envFiles: envFiles
            });
        })
        : file_utils_1.Right(config);
};
/**
 * Either wrapper to handle the fact that 1 or more of the envFile
 * options could've resulted in an error.
 * If that's the case, we want to return a Left that has the report type
 * `EnvFileOptionError` so that a nice error can be printed
 */
exports.getEnvFiles = function (envFile, globalPrefix, cwd) {
    return file_utils_1.Right([].concat(envFile).map(function (path) { return exports.getSingleEnvFile(path, globalPrefix, cwd); }))
        .map(function (xs) {
        return {
            all: xs,
            valid: xs.filter(function (x) { return x.errors.length === 0; }),
            invalid: xs.filter(function (x) { return x.errors.length > 0; })
        };
    })
        .chain(function (x) { return x.invalid.length
        ? file_utils_1.Left({ type: reports.ReportTypes.EnvFileOptionError, data: x })
        : file_utils_1.Right(x.valid); });
};
/**
 * Given a single `envFile` option, try to read & parse
 * a file from disk.
 */
exports.getSingleEnvFile = function (envFile, globalPrefix, cwd) {
    return file_utils_1.Right(envFile)
        // Get an array of 'lookupPath' and 'prefix' in the same scope
        .chain(function (envFileItem) { return file_utils_1.Right([getLookupPath(envFileItem), getPrefix(envFileItem, globalPrefix)])
        // destructure previous results
        .chain(function (_a) {
        var lookupPath = _a[0], prefix = _a[1];
        // Now attempt to read the file from disk
        return readEnvFile(lookupPath, cwd)
            // If an error occurs here, delegate to a `EnvFileNotFound` error
            .fold(function (file) {
            return file_utils_1.Left({
                input: lookupPath,
                file: file,
                prefix: prefix,
                errors: [{ type: task_utils_1.InputErrorTypes.EnvFileNotFound }]
            });
        }, function (r) { return file_utils_1.Right(r); }) // <-- otherwise continue with the result
            // Now we have a file successfully read from disk, keep it in scope
            .chain(function (file) {
            // Try to parse the data from it. (JSON/env could throw)
            return file_utils_1.tryCatch(function () { return parseData(file); })
                // If an error occurs in parsing, defer to `EnvFileParseError` error
                .fold(function (error) { return file_utils_1.Left({
                input: lookupPath,
                file: file,
                prefix: prefix,
                errors: [{ type: task_utils_1.InputErrorTypes.EnvFileParseError, error: error }]
            }); }, 
            // If the content parsed without error, return the final result
            function (parsedData) { return file_utils_1.Right({
                input: lookupPath,
                file: _.assign({}, file, { data: parsedData }),
                prefix: prefix, errors: []
            }); });
        });
    }); }).fold(function (e) { return e; }, function (r) { return r; }); // finally fold as we could've exited early todo: could this be removed?
};
/**
 * Wrapper for reading the env file from disk.
 * The result could have an 'errors' property so we
 * wrap this in the Either type
 */
var readEnvFile = function (lookupPath, cwd) {
    return file_utils_1.Right(file_utils_1.readFileFromDiskWithContent(lookupPath, cwd))
        .chain(function (result) { return result.errors.length
        ? file_utils_1.Left(result)
        : file_utils_1.Right(result); });
};
/**
 * Parse either json or regular .env file
 * @param file
 * @returns {any}
 */
function parseData(file) {
    if (file.parsed.ext === '.json') {
        return JSON.parse(file.content);
    }
    return file_utils_1.parseEnv(file.content);
}
/**
 * `envFile` could be either a string or object, so
 * we need to decide where to get the lookup path from
 *
 * -> envFile: package.json
 * -> envFile:
 *      path: package.json
 *      prefix: [npm, package]
 */
var getLookupPath = function (envFile) {
    return typeof envFile === 'string'
        ? envFile
        : envFile.path;
};
/**
 * envFiles can either have their own prefixes array, or
 * it will inherit from the global if not set.
 * @param envFile
 * @param globalPrefixes
 * @returns {any}
 */
var getPrefix = function (envFile, globalPrefixes) {
    if (typeof envFile === 'string') {
        return globalPrefixes;
    }
    if (envFile.prefix && envFile.prefix.length) {
        return [].concat(envFile.prefix);
    }
    return globalPrefixes;
};
