// crossbow/dist/task.sequence.factories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("../lodash.custom");
var SequenceItemTypes;
(function (SequenceItemTypes) {
    SequenceItemTypes[SequenceItemTypes["SeriesGroup"] = "SeriesGroup"] = "SeriesGroup";
    SequenceItemTypes[SequenceItemTypes["ParallelGroup"] = "ParallelGroup"] = "ParallelGroup";
    SequenceItemTypes[SequenceItemTypes["Task"] = "Task"] = "Task";
})(SequenceItemTypes = exports.SequenceItemTypes || (exports.SequenceItemTypes = {}));
var seqUID = 0;
function createSequenceTaskItem(incoming) {
    return _.assign({ type: SequenceItemTypes.Task, items: [], seqUID: seqUID++ }, incoming);
}
exports.createSequenceTaskItem = createSequenceTaskItem;
function createSequenceSeriesGroup(incoming) {
    return _.assign({ type: SequenceItemTypes.SeriesGroup }, incoming);
}
exports.createSequenceSeriesGroup = createSequenceSeriesGroup;
function createSequenceParallelGroup(incoming) {
    return _.assign({ type: SequenceItemTypes.ParallelGroup }, incoming);
}
exports.createSequenceParallelGroup = createSequenceParallelGroup;
